library/std/src/env.rs

```diff
@@ -318,20 +318,18 @@ impl Error for VarError {
 ///
 /// # Safety
 ///
-/// Even though this function is currently not marked as `unsafe`, it needs to
-/// be because invoking it can cause undefined behaviour. The function will be
-/// marked `unsafe` in a future version of Rust. This is tracked in
-/// [rust#27970](https://github.com/rust-lang/rust/issues/27970).
-///
 /// This function is safe to call in a single-threaded program.
 ///
-/// In multi-threaded programs, you must ensure that are no other threads
-/// concurrently writing or *reading*(!) from the environment through functions
-/// other than the ones in this module. You are responsible for figuring out
-/// how to achieve this, but we strongly suggest not using `set_var` or
-/// `remove_var` in multi-threaded programs at all.
+/// This function is also always safe to call on Windows, in single-threaded
+/// and multi-threaded programs.
 ///
-/// Most C libraries, including libc itself do not advertise which functions
+/// In multi-threaded programs on other operating systems, you must ensure that
+/// are no other threads concurrently writing or *reading*(!) from the
+/// environment through functions other than the ones in this module. You are
+/// responsible for figuring out how to achieve this, but we strongly suggest
```

Comment:
So it's already the caller responsibility to avoid racing, and the functions are unsafe now. Then could we remove the lock in the implementation of `_set_var` and `_remove_var`?

Reply:
Let's not overdo it. There's likely tons of Rust code out there that isn't super careful about environment variable races, and that's entirely our fault as we made these functions safe. I don't think there is a reason to make this code more broken than it is today.

Maybe once a large fraction of the ecosystem is adjusted to the new world, we can consider this. But the lock doesn't cost us that much (IIRC even the fork/exec in `Command` now deals with the lock properly), so I feel like it'd be a disservice to all the code out there that relies on the lock to remove it "just because we can".